### PR TITLE
Feature/middle click open in new independent window

### LINKED ISSUE #1886: [Windows 7, 8 & 8.1] 0.15.1 crashes upon video playback (very beginning, reproducable)
**Behavior of the program**

- Freetube crashes as soon as any video is played 
- it goes to the player environment, starts the playback 
- image freezes at 0:00 or 0:01 of the video and program is reported as not responding
- sound is playing for 2-10 seconds while Freetube is reported as unresponsive, then stops

**Expected behavior**
Video playback without Freetube crashing and being unresponsive.

**To Reproduce**
Please add all steps to reproduce the behavior:
1. Go to 'any video (for example https://youtu.be/mefTPTD-p6k)
2. wait for video to load and begin playing
3. See error

**Screenshots/DevConLogs**
DevTools failed to load source map: Could not parse content for file:///C:/Program%20Files/FreeTube/resources/app.asar/dist/index.js.map: Unexpected end of JSON input

Bild_2021-11-12_011403

**Environment Information (please complete the following information):**
 - OS: Windows
 - OS Version: Windows 7 SP 1 all patches
 - FreeTube version: 0.15.1 beta
 - Installation Method (If applicable): Win-x64 installer .exe
 - Primary API Used: Local API, also tried with Invidious API

### LINKED ISSUE #1957: instances more than 1 child deep don't work
There was an issue where you could only watch videos in the main parent window, opening children windows would not load anything.
That has been corrected in a recent update, but still it works with a single level as depth. Opening a new instance/window from that child is still not working.
With earlier versions there wasn't this limitation, and it's really annoying that I can't follow new videos from child instances. For example I'm looking at a how-to video which has interesting videos in the playlist. I can't open a new one while keeping the existing one open as well. I mean I can, but it's not loading anything. I have to abandon the video I was watching and that it's pretty annoying since I can't correlate info between videos this way. Main parent window is usually where I keep my subscription list that I keep refreshing once in a while. That means that I'm stuck with the first child instances and I have to see every video in them. 
A workaround is to open a new child from the parent and navigate in it to the other video I want to have on screen at the same time with the previous child.
Is it that complicated to allow for 2-3 levels deep?

### PR DESCRIPTION
**Pull request by PikachuEXE**
**Important note**
We may remove your pull request if you do not use this provided PR template correctly.

**Pull Request Type**
Please select what type of pull request this is:
- [x] Bugfix

**Related issue**
Closes #2006 since this PR also includes changes for that PR
Closes #1957
Closes #1886

**Description**
Existing behaviour for opening new window via interaction like middle mouse button click on video/channel links
- Link clicked (let's call the window with the clicked link as "parent window")
- A new window is created (let's call it "child window")
- Closes "parent window"
- "child window" **automatically** closed

**Screenshots (if appropriate)**
Before change:

https://user-images.githubusercontent.com/1018543/141750691-e3483e52-c994-449f-853d-7da9e9855b8f.mov

After change (taken in dev, prod still require testing):

https://user-images.githubusercontent.com/1018543/142129914-76d36850-d7da-4676-bccf-b27d39478f22.mov

**Testing (for code that is not small enough to be easily understandable)**
**TWO** parts - development and production
Common for both dev and prod:
- Link clicked (let's call the window with the clicked link as "parent window")
- A new window is created (let's call it "child window")
- Closes "parent window"

Expected "child window" not closed

Dev: `npm run dev` / `npm run debug`
Prod: `npm run build --if-present`, run the app

**Desktop (please complete the following information):**
 - OS: MacOS
 - OS Version: 11.6.1
 - FreeTube version: based on ab918383d4365ef1743fb80a97c5bc0115c9529d

**Additional context**
Still got some issues
- [x] New window always maximized (No need fix - seems to be just window size for new window not remembered yet)

Like to have code conflict with current #1833

### CODE REVIEW

**Comment by peepopoggers** [comment]
lol

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
I tried to middle click the channels in my sidebar but that doesnt do anything

I thought that would also work because it works when u click on a channel name beneath the video title

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
Tried to click a channel link in the description but that also didnt work

https://www.youtube.com/watch?v=N9PcNrhiMUc

In the comment somebody added a timestamp and if i middle click it will open a window on the subscription page

**Comment by peepopoggers** [comment]
Is this a mac only issue? does not happen on linux for me

**Review by efb4f5ff-1298-471a-8973-3d47447115dc — DISMISSED** [review]
Works for me tested on Ubuntu

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
> Is this a mac only issue? does not happen on linux for me

No, this also occurred on windows

**Comment by PikachuEXE** [comment]
> No, this also occurred on windows
I use FT on both Windows and MacOS
And this is the current mostly annoying issue for me
I workaround by copying the video URL, open new window and paste on search bar (even **more** annoying)

**Comment by peepopoggers** [comment]
> Works for me tested on Ubuntu

Does it happen before to you? I do not have this behavior on my machine.

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
Yes on ubuntu and windows this happened to me before

**Comment by PikachuEXE** [comment]
I make GH run build workflow once
https://github.com/FreeTubeApp/FreeTube/actions/runs/1469729377

I will test the windows version tonight

Oh wait maybe I can try some VM...

**Comment by PikachuEXE** [comment]
Spotted a delay in new window being shown
Updated code to fix that

https://user-images.githubusercontent.com/1018543/142129914-76d36850-d7da-4676-bccf-b27d39478f22.mov

**Comment by PikachuEXE** [comment]
Latest build:
https://github.com/FreeTubeApp/FreeTube/actions/runs/1469918797

**Comment by PikachuEXE** [comment]
Please start testing this soon~ (especially Linux coz I don't use)
I am currently testing on Windows & MacOS by using it as daily FT

**Comment by PikachuEXE** [comment]
I have also made a change to make new window created by new window button appear sooner
Mind if I put that change into this PR too?

https://user-images.githubusercontent.com/1018543/142378799-33072070-f01d-4ae5-8b13-82bd2728dd7f.mov

**Comment by PikachuEXE** [comment]
Do you have more feedback for this? @peepopoggers

The only issue I got is if a window is shown immediately then [the flash](https://github.com/FreeTubeApp/FreeTube/pull/1502) is back

**Comment by peepopoggers** [comment]
Not the flash :sob: What does the code look like for opening a duplicate window?

**Comment by PikachuEXE** [comment]
Problem is the code did set background color for the window
https://github.com/PikachuEXE/FreeTube/blob/feature/middle-click-open-in-new-independent-window/src/main/index.js#L183-L207
Unless there are some options I don't know that also need to be set
It should be Electron issue?

**Comment by PikachuEXE** [comment]
Update to only focus on new window if not already shown

**Comment by PikachuEXE** [comment]
@peepopoggers 
Just did some testing with a minimal app
https://github.com/PikachuEXE/my-electron-app

Which shows upgrading electron to 16.x fixes the flash on new window...
ELECTRON again!?

**Comment by PikachuEXE** [comment]
This is my locally built app with electron 16

https://user-images.githubusercontent.com/1018543/143532350-295b77c9-34fc-45a2-803b-110bb68411d2.mov

**Comment by ChunkyProgrammer** [comment]
@PikachuEXE github saves edit history so I removed your comment for you.

PikachuEXE: 
> This is my locally built app with electron 16

https://user-images.githubusercontent.com/1018543/143532576-36f6f3f2-4dd7-4a5e-a21b-a9243122fff8.mov


> Edit: Updated video to avoid exposing unwanted stuff :P

**Comment by PikachuEXE** [comment]
Oh ya forgot that thx

**Comment by PikachuEXE** [comment]
Created #1925 to upgrade electron first

**Comment by PikachuEXE** [comment]
Code rebased on `development` with #1833 merged with [test builds available](https://github.com/PikachuEXE/FreeTube/actions/workflows/build.yml)

**Review by efb4f5ff-1298-471a-8973-3d47447115dc — DISMISSED** [review]
LGTM!

**Review by ChunkyProgrammer — DISMISSED** [review]
LGTM. #1925 should probably be closed

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
This PR closes that PR so it indeed doesn't make much sense in keeping that one open if this one also fixes it. I'll close it.

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
ready for review right?

**Comment by rhinoduck** [comment]
As per electron's release notes, the media playback crashes have been fixed in [v16.0.6](https://github.com/electron/electron/releases/tag/v16.0.6) so this should indeed close https://github.com/FreeTubeApp/FreeTube/issues/1886 too.

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
Closes #1957

**Comment by PikachuEXE** [comment]
#1923 requires a separate fix
But this PR is not intended to fix anything like that

https://github.com/FreeTubeApp/FreeTube/commit/2be2301d92e37d13a4839ded1fc436836292b7d4 fixes it partially maybe?

I will add #1886 & #1957

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
last question before approval

if i middle click on a invidious url would it also has the same behavior like middle clicking the youtube url?

**Review by efb4f5ff-1298-471a-8973-3d47447115dc — APPROVED** [review]

**Comment by PikachuEXE** [comment]
@peepopoggers @constraintAutomaton 
Would you have time to review this PR? 
Some stuff are already depending/conflicting with this PR:
- #2027 
- Complete fix for #1809 (there is a fix but only open link in same window even when middle click)

**Comment by PikachuEXE** [comment]
`electron` updated to [`16.0.8`](https://github.com/electron/electron/releases/tag/v16.0.8)

**Review by efb4f5ff-1298-471a-8973-3d47447115dc — DISMISSED** [review]
Damn, Electron moving fast with those releases.

Nothing breaks so LGTM!

File: src/main/index.js [hunk 1]
```diff
@@ -171,7 +171,7 @@ function runApp() {
     }
   }
 
-  async function createWindow(replaceMainWindow = true) {
+  async function createWindow({ replaceMainWindow = true, windowStartupUrl = null, showWindowNow = false } = { replaceMainWindow: true, windowStartupUrl: null, showWindowNow: false }) {
```
Comment (ChunkyProgrammer) [inline src/main/index.js:174]: You can avoid code repetition by doing this
```suggestion
  async function createWindow({ replaceMainWindow = true, windowStartupUrl = null, showWindowNow = false } = { }) {
```

**Comment by PikachuEXE** [comment]
Updated~

**Review by efb4f5ff-1298-471a-8973-3d47447115dc — APPROVED** [review]
👍⚡

File: src/main/index.js [hunk 2]
```diff
@@ -257,10 +255,18 @@ function runApp() {
 
     // load root file/url
     if (isDev) {
-      newWindow.loadURL('http://localhost:9080')
+      let devStartupURL = 'http://localhost:9080'
+      if (windowStartupUrl != null) {
```
Comment (PrestonN) [inline src/main/index.js:259]: `!=` should be `!==`

File: src/main/index.js [hunk 3]
```diff
@@ -257,10 +255,18 @@ function runApp() {
...
+        devStartupURL = windowStartupUrl
+      }
+      newWindow.loadURL(devStartupURL)
     } else {
-      /* eslint-disable-next-line */
-      newWindow.loadFile(`${__dirname}/index.html`)
+      if (windowStartupUrl != null) {
```
Comment (PrestonN) [inline src/main/index.js:264]: Same thing here

**Review by PrestonN — APPROVED** [review]
LGTM

**Pull request merged**